clients/apple/Shared/Stateful Logic/BillingService.swift

```diff
@@ -0,0 +1,198 @@
+import SwiftLockbookCore
+import SwiftUI
+import StoreKit
+
+public enum StoreError: Error {
+    case failedVerification
+    case noProduct
+}
+
+public enum PurchaseResult {
+    case success
+    case pending
+    case failure
+}
+
+public enum CancelSubscriptionResult {
+    case success
+    case appstoreActionRequired
+}
+
+
+let MONTHLY_SUBSCRIPTION_PRODUCT_ID = "basic.premium"
+
+class BillingService: ObservableObject {
+    let core: LockbookApi
+    
+    var pendingTransactionsListener: Task<Void, Error>? = nil
+
+    var maybeMonthlySubscription: Product? = nil
+    var purchaseResult: PurchaseResult? = nil
+    @Published var cancelSubscriptionResult: CancelSubscriptionResult? = nil
+    
+    var makingPurchaseAttempt = false
+    
+    
+    init(_ core: LockbookApi) {
+        self.core = core
+        pendingTransactionsListener = listenForTransactions()
+
+        Task {
+            await requestProducts()
+        }
+    }
+    
+    deinit {
+        pendingTransactionsListener?.cancel()
+    }
+    
+    func listenForTransactions() -> Task<Void, Error> {
+        return Task.detached {
+            for await verification in Transaction.updates {
+                do {
+                    let transaction = try self.checkVerified(verification)
+                    switch self.core.getUsage() {
+                    case .success(let usages):
+                        if usages.dataCap.exact == 1000000 {
+                            if let receipt = self.getReceipt(), !self.makingPurchaseAttempt, transaction.id == transaction.originalID {
+                                guard let appAccountToken = transaction.appAccountToken else {
+                                    throw StoreError.failedVerification
+                                }
+                                
+                                let result = self.core.newAppleSub(originalTransactionId: String(transaction.originalID), appAccountToken: appAccountToken.uuidString.lowercased(), encodedReceipt: receipt)
+                                
+                                switch result {
+                                case .success(_):
+                                    await transaction.finish()
+                                    self.purchaseResult = .success
+                                case .failure(let error):
+                                    DI.errors.handleError(error)
+                                }
+                            }
+                        } else {
+                            await transaction.finish()
+                        }
+                    case .failure(let error):
+                        DI.errors.handleError(error)
+                    }
+                } catch {
+                    print("Transaction failed verification")
```

Comment:
This is never going to make it to the user, we have a way to handle unexpected errors